[BEAM-461] Create custom Kinesis source

ppastuszkaocado opened the pull request:
This adds custom Kinesis source as new beam-sdks-java-io-kinesis module.
Build on Travis has passed.

This has been tested with unit tests (included in the pull request), but also using DirectRunner (I've confirmed, that source is able to successfully read data produced by Kinesis Producer Library). I've also tried to run this on real dataflow, but it doesn't seem to work (the pipeline starts, but it doesn't read anything at all) - I'm going to investigate the issue, but I'd like you to do the review in the meantime.
Speaking about tests... I've got bunch of integration / end-to-end level tests (using DirectRunner and DataflowRunner), which are using real kinesis to read data from. I'm not sure where should I put them and how would they be executed in apache CI environment?

BTW. Is this "large" enough, so that I should sign ICLA?

@dhalperi, @jbonofre - can you please review this?

ppastuszkaocado commented:
This have been now successfully tested on real dataflow pipeline (previously it wasn't working when I tried to use the runner code from master, but worked smoothly when I fell back to 0.1.0-incubating).

dhalperi commented:
R: @dhalperi

dhalperi commented on sdks/java/io/kinesis/pom.xml:
```diff
@@ -0,0 +1,141 @@
+    <dependency>
+      <groupId>com.amazonaws</groupId>
+      <artifactId>aws-java-sdk-core</artifactId>
+      <version>1.11.18</version>
```
I see this version number is used twice. Should you extract a property `aws.sdk.version`and use it twice to keep them in sync?

dhalperi replied:
Is the AWS SDK on the `KinesisIO` API surface or is it only an internal dependency?

Ideally, it is not on the public API surface and you can shade it.

But if not, we should move it to the top-level pom under `dependencyManagement` so that all modules will use the same version.

ppastuszkaocado replied:
As a disclaimer: aws sdk is a transition dependency of amazon kinesis client. Ideally, I wouldn't want to include that explicitly in the pom, but if I don't do this, then maven-dependency-plugin complains about usage of undeclared dependencies.
What should I do with this then?

dhalperi replied:
Hi @ppastuszkaocado 

Looking through this again. `aws-sdk-java-core` is actually used explicitly by you, e.g., you reference [`Regions`](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-core/src/main/java/com/amazonaws/regions/Regions.java) in your code, which is part of the `aws-java-sdk-core` code.

The Maven enforcer makes sure that you directly include in your pom those transitive dependencies that you actually use directly. This way, if Kinesis somehow drops the transitive dependency on AWS, your code will still compile. (I know this is unlikely, but it's a safe policy).

dhalperi commented on sdks/java/io/kinesis/src/main/java/com/google/common/base/AbsentWithNoSuchElement.java:
```diff
@@ -0,0 +1,92 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.common.base;
```
please do not include code in other packages than `KinesisIO`. We can just shade Guava, if that's what you're trying to do.

ppastuszkaocado replied:
Done.

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -0,0 +1,117 @@
+ * {@link PTransform}s for reading from
+ * <a href="https://aws.amazon.com/kinesis/">Kinesis</a> streams.
+ */
+public class KinesisIO {
```
`final`?

dhalperi replied:
Usually, `KinesisIO` class-level Javadoc is where we have the copious Javadoc information that would say how this connector is used. This is the information you put in the `README.md` -- how to construct the readers, etc. It's important to have here since most people will be using IDEs to edit and not looking at github :).

This is also where you would talk about timestamps -- in this case, that records are always given the processing time as their timestamp.

See KafkaIO.java for an example?

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/client/SimplifiedKinesisClient.java:
```diff
@@ -0,0 +1,164 @@
+ * Wraps {@link AmazonKinesis} class providing much simpler interface and
+ * proper error handling.
+ */
+public class SimplifiedKinesisClient {
```
can these and other utility classes that are not on the public API surface of `KinesisIO` be package-private?

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/client/response/GetKinesisRecordsResult.java:
```diff
@@ -0,0 +1,54 @@
+import javax.annotation.Nullable;
+
+/***
+ *
```
Javadoc?

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/client/response/KinesisRecord.java:
```diff
@@ -0,0 +1,90 @@
+     * @return unique id of the record based on its position in the stream
+     */
+    public byte[] getUniqueId() {
+        return getExtendedSequenceNumber().toString().getBytes(Charsets.UTF_8);
```
[`StandardCharsets.UTF_8`](https://docs.oracle.com/javase/7/docs/api/java/nio/charset/StandardCharsets.html#UTF_8) -- no need to use Guava just for this :)

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/source/KinesisRecordCoder.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.List;
+
+/**
+ * Created by p.pastuszka on 07/04/16.
```
General javadoc request for details. See, e.g., AvroCoder or similar.

If this is a package-private coder, then you can be much more straightforward e.g., `A {@link Coder} for {@link KinesisRecord}`.

dhalperi commented on sdks/java/io/kinesis/pom.xml:
```diff
@@ -0,0 +1,141 @@
+    <dependency>
+      <groupId>org.mockito</groupId>
+      <artifactId>mockito-all</artifactId>
+      <version>1.10.19</version>
```
do you need to override the version from parent? If so, add a comment?

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -0,0 +1,117 @@
+    public static class Read {
+
+        private final String streamName;
+        private final StartingPoint initialPosition;
```
can these be null? If so, `@Nullable`. If not, use input validation

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/client/SimplifiedKinesisClient.java:
```diff
@@ -0,0 +1,164 @@
+            throw new RuntimeException(e);
+        } catch (Exception e) {
+            LOG.error("Unknown failure", e);
+            throw new RuntimeException(e);
```
Can you include the log messages in the created exceptions?

(Also, maybe, maybe the caller could choose whether to log.)

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/client/response/KinesisRecord.java:
```diff
@@ -0,0 +1,90 @@
+ * {@link UserRecord} enhanced with utility methods.
+ */
+public class KinesisRecord extends UserRecord {
+    private Date readTime = new Date();
```
Could you use joda Instant instead? More like the rest of the SDK.

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/source/KinesisReader.java:
```diff
@@ -0,0 +1,158 @@
+            for (int i = 0; i < shardIterators.size(); ++i) {
+                currentRecord = shardIterators.getCurrent().next();
+                if (currentRecord.isPresent()) {
+                    currentRecordTimestamp = Optional.of(Instant.now());
```
Why not `currentRecord.getReadTime()` or whatever it would be.

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -0,0 +1,117 @@
+            return new Read(streamName, new StartingPoint(initialPosition));
+        }
+
+        public static Read from(String streamName, Date initialTimestamp) {
```
Date -> Instant?

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/source/KinesisRecordCoder.java:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Created by p.pastuszka on 07/04/16.
+ */
+public class KinesisRecordCoder extends StandardCoder<KinesisRecord> {
```
I think this can be implemented simpler as an `AtomicCoder` -- you need not register components since you don't have any.

dhalperi replied:
Use `CoderProperties` to test. See, e.g., [`MapCoderTest`](https://github.com/apache/incubator-beam/blob/69e22023f7efe3a216776297293e9d9c34d8e534/sdks/java/core/src/test/java/org/apache/beam/sdk/coders/MapCoderTest.java)

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/source/KinesisRecordCoder.java:
```diff
@@ -0,0 +1,96 @@
+    private static final StringUtf8Coder stringCoder = StringUtf8Coder.of();
+    private static final ByteArrayCoder byteArrayCoder = ByteArrayCoder.of();
+    private static final InstantCoder instantCoder = InstantCoder.of();
+    private static final VarLongCoder varLongCoder = VarLongCoder.of();
```
naming: `static final` member variables should be named like `STRING_CODER`

dhalperi commented:
I saw your question about tests.

For unit tests, I bet you can either mock or fake out the Kinesis library. That way we can test the code without relying on the service.

For integration tests: I'm assuming that there's no "free, public" Kinesis installation anywhere nor a single-machine version we can run? For KafkaIO, we run a local Kafka instance during unit tests. For several other services, we have an Apache Beam account where we have set it up specifically for testing. We may need to set up something similar for continuous integration testing of Kinesis. We'll have to figure out how we could get accounts, etc. with the right Apache control to do so.

As far as writing an end-to-end test.. If you look at [BigtableIO tests, e.g., BigtableReadIT](https://github.com/apache/incubator-beam/blob/master/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableReadIT.java) you'll see how those are set up. They are setup as Surefire integration tests (disabled by default) and they typically rely on pipeline options to be set in the environment variables + manually turning them on in order to run. In this manner, someone with access to a Kinesis installation can run the test by setting up the right test pipeline options.

In terms of things we need to do before merging .. someone can run your integration test manually and defer Jenkins-ifying it until later.

dhalperi commented:
When you say "run this on real Dataflow" can you say more?

* Beam HEAD with the Google Cloud Dataflow runner?
* Google Cloud Dataflow SDK for Java 1.6.0?

something else?

More info, I can look into it.

dhalperi commented:
Finally, I wanted to say thanks for porting forward from your PR against the Dataflow SDK to the Beam SDK here. It's really awesome to see how this has progressed from [our original discussion on StackOverflow](http://stackoverflow.com/questions/34118103/how-custom-unbounded-sources-work-in-google-cloud-dataflow/34164264#34164264) :)

ppastuszkaocado commented:
Thank you for review and comments, @dhalperi. I have addressed most of the issues in new commit.

About tests:
I will add a unit test, which mocks AmazonKinesis and runs whole job using direct pipeline runner. This should give us more confidence in the solution.

I'm not aware of any "free" full Kinesis implementation (I've found in-memory kinesis implementation: https://github.com/mhart/kinesalite, but I'm not sure it'll give us more than mocking the AmazonKinesis interface in unit tests). Moreover, Kinesis has no free quota in Amazon, so you need to have billing enabled to use it :(
I took a look at BigTableReadIT test and it seems like I can do a similar thing for Kinesis - then, as you said, someone would need to provide kinesis option to run this and it'll have to be manually enabled.

When I said "run this on real Dataflow" I meant Beam 0.1.0-incubating on Google Cloud Dataflow.

P.S. I can send you my current integration / e2e tests if you'd like to.

ppastuszkaocado commented:
Hi @dhalperi,

I've added KinesisMockReadTest, which mocks kinesis and runs using directrunner and very similar integration test called KinesisReaderIT - this one uses the real kinesis.

Can you please take a look now?

jbonofre commented:
R: @jbonofre

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/CustomOptional.java:
```diff
@@ -0,0 +1,87 @@
+import java.util.NoSuchElementException;
+
+/***
+ * Similar to {@link Optional}, but throws {@link NoSuchElementException} for missing element.
```
Nit: would prefer something like `Similar to Guava {@code Optional}` -- because we are bringing in an import on Guava just for the javadoc.

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -34,44 +33,44 @@
+ * <p>
  * <h3>Usage</h3>
- *
+ * <p>
```
double `<p>` necessary? here and below.

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -34,44 +33,44 @@
+ * <ul>
+ * <li>credentials (aws key, aws secret)</li>
+ * <li>region where the stream is located</li>
+ * </ul>
```
the indent here was probably helpful. But fine as is.

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -164,6 +162,10 @@ public static Read from(String streamName, Instant initialTimestamp) {
             private final Regions region;
 
             private BasicKinesisProvider(String accessKey, String secretKey, Regions region) {
+                checkNotNull(accessKey);
```
prefer something like

```java
checkNotNull(accessKey, "accessKey");
checkNotNull(secretKey, "secretKey");
...
```

so that users get a helpful error message.

(I lean away from "A secret key must be provided" because it's unnecessary IMO.)

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -164,6 +162,10 @@ public static Read from(String streamName, Instant initialTimestamp) {
+                checkNotNull(region);
+
                 this.accessKey = accessKey;
                 this.secretKey = secretKey;
```
can actually inline the null checks, also:

```java
this.accessKey = checkNotNull(accessKey, "accessKey");
```

dhalperi commented on sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisReaderIT.java:
```diff
@@ -0,0 +1,117 @@
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import org.junit.Test;
+import static java.util.concurrent.Executors.newSingleThreadExecutor;
```
probably checkstyle will complain that `static` imports are not pulled out separately at top.

dhalperi commented on sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisReaderIT.java:
```diff
@@ -0,0 +1,117 @@
+        @Override
+        public void processElement(ProcessContext c) throws Exception {
+            checkNotNull(c.element());
+            c.output(new String(c.element().getData().array(), Charsets.UTF_8));
```
`StandardCharsets`?

dhalperi commented on sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisUploader.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.List;
+
+/**
+ *
```
Javadoc?

dhalperi commented on sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisUploader.java:
```diff
@@ -0,0 +1,82 @@
+                                options.getAwsAccessKey(), options.getAwsSecretKey()))
+        ).withRegion(Regions.fromName(options.getAwsKinesisRegion()));
+
+        List<List<String>> partitions = Lists.partition(data, 499);
```
where does 499 come from? Should this also be a test parameter?

dhalperi commented on sdks/java/io/kinesis/src/test/java/org/apache/beam/sdk/io/kinesis/KinesisReaderIT.java:
```diff
@@ -0,0 +1,117 @@
+    private ExecutorService singleThreadExecutor = newSingleThreadExecutor();
+
+
+    @Test
```
Jenkins is failing on this test. Can you please `@Ignore` this and file a JIRA issue in the `testing` component for us to set up a Kinesis instance?

ppastuszkaocado commented:
@dhalperi - I've addressed your comments. Please, take a look now.
I've create appropriate issue in JIRA for Kinesis IT tests once this is merged.

ppastuszkaocado commented:
@dhalperi - any progress on this?

dhalperi commented on sdks/java/io/kinesis/pom.xml:
```diff
@@ -0,0 +1,179 @@
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>beam-runners-direct-java</artifactId>
+      <version>0.3.0-incubating-SNAPSHOT</version>
```
drop

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisReader.java:
```diff
@@ -0,0 +1,146 @@
+                }
+            }
+        } catch (TransientKinesisException e) {
+            LOG.warn("Transient exception occurred: {}", e);
```
no `{}` necessary on logs like this: slf4j will automatically recognize exception class and do the right thing

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/KinesisIO.java:
```diff
@@ -0,0 +1,186 @@
+ * As you can see you need to provide 3 things:
+ * <ul>
+ *   <li>name of the stream you're going to read</li>
+ *   <li>position in the stream where reading should start. There are two options:</li>
```
does this render right? I'd think the `<ul>` should be inside this `<li>`

ppastuszkaocado replied:
Yes, it does render right: http://scratchpad.io/sweet-rat-5644

dhalperi commented on sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/SimplifiedKinesisClient.java:
```diff
@@ -0,0 +1,159 @@
+    /***
+     * Wraps Amazon specific exceptions into more friendly format.
+     *
+     * @throws IOException              - in case of recoverable situation, i.e.
```
This does not seem to match the implementation.

have you changed `IOException` out for `TransientKinesisException`?

dhalperi commented:
Minor nits, but LGTM at this point.

In addition to nits, if you have time I'd appreciate two quick passes:

* do a minor review of Javadoc everywhere. I still see some empty blocks for top-level classes (`UnboundedSource` eg)
* throw useful errors. Overall good, but there are still lots of `checkNotNull(f)` which will not throw the most useful exception.

Let me know when you're ready for me to push. The only "mandatory" comments are the ones in the pom.xml

dhalperi commented:
I made a few fixups, and merged. Thanks @ppastuszkaocado !
